### PR TITLE
Encapsulate std::complex functions

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

For the last few releases Stan math has had to do quick fixes when clang would update their functions for 'std::complex' which can cause conflicts when looking for our own functions that work over complex. The main issue for function calls like the following where an unqualified name lookup is used to find the correct function to call.

```c++
struct sin_fun {
  template <typename T>
  static inline auto fun(const T& x) {
    using std::sin;
    return sin(x);
  }
};
```

In the above, `sin` is looked for in both the `stan::math` namespace and the `std` namespace since [unqualified name lookup](https://en.cppreference.com/w/cpp/language/unqualified_lookup) also uses [Argument Dependent Lookup](https://en.cppreference.com/w/cpp/language/adl) (ADL). During ADL the compiler looks at the namespaces of the arguments as well as the local namespace to figure out all of the valid function signatures it should use when deciding which function is the best candidate. Since we override `std::complex` the `std` namespace is then also included in the valid candidates. Sometimes this is fine because the clang standard library uses `requires` very similar to ours for only supporting floating point types inside of their complex functions.

```c++
template <class _Tp, __enable_if_t<is_floating_point<_Tp>::value, int> >
_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 complex<_Tp>
operator*(const complex<_Tp>& __lhs, const complex<_Tp>& __rhs) {
  return complex<_Tp>(__from_builtin_tag(), __lhs.__builtin() * __rhs.__builtin());
}
```

But some functions, like the one below (linked [here](https://github.com/llvm/llvm-project/blob/main/libcxx/include/complex#L984)), do not restrict themselves to only floating point types.

```c++
template <class _Tp>
inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 complex<_Tp> conj(const complex<_Tp>& __c) {
  return complex<_Tp>(__c.real(), -__c.imag());
}
```

This causes issues because ADL then finds two valid candidates functions when trying to find the best candidate for `conj` (both the `stan::math` version and the `std` version).

There is no good solution to this issue, it's just how the language is defined. The only way to make `stan::math` functions a better candidate is to use templates to make the stan math function both more generic and have an extra template with our `requires` which makes it slightly better so the compiler chooses it as the best candidate.

So this PR does a few things.

1. Use templates for all of the std functions that we want to override for complex types. The generic template and parameters makes our functions a better candidate. 
    - This is very much a patch solution. In the long run we need our own custom complex scalar type. But we can't do that yet till an issue is resolved in eigen (see [here](https://gitlab.com/libeigen/eigen/-/issues/2865)) where they assume users are overriding `std::complex` for their complex types.
2. Cleans up the vectorized functions signatures. There were 4 or 5 different requires being used and they made the function rather confusing. Most of the vectorized functions work over containers containing base types of `autodiff` so we can just use one requires instead of the several that were there
3. Fixes the headers in rev so that `rev` headers are always used first and if available. This stops issues with prim being brought in before their respective `rev` or `fwd` functions are available and fixes soem header include order issues.

## Tests

No changes to tests

## Side Effects

Hopefully not

## Release notes

Template complex functions in the Stan math library to be better candidates than the standard library functions during ADL.

## Checklist

- [x] Copyright holder: Simons Foundation

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
Also after this release I'm going to write a script that we can use in jenkins to check that includes are being done correctly

**Comment by WardBrian** [comment]
Tagging #3006 just so this PR shows up in the history of that issue -- as you say, actually resolving it would require Eigen to make some changes first

**Comment by WardBrian** [comment]
Running with latest compilers: https://jenkins.flatironinstitute.org/blue/organizations/jenkins/Stan%2FBleedingEdgeCompilersMonthly/detail/BleedingEdgeCompilersMonthly/235/pipeline

**Review by WardBrian — CHANGES_REQUESTED** [review]
A bunch of tiny questions/comments -- thanks for tackling this!

File: stan/math/fwd/fun/conj.hpp
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/fwd/fun/conj.hpp:1]: Yes I'm going to write that script to make a file for each file in prim soon for fwd and rev so it's just easier to leave this here for now

File: stan/math/fwd/fun/atan2.hpp [hunk 1]
```diff
@@ -3,30 +3,27 @@
 
 #include <stan/math/fwd/meta.hpp>
 #include <stan/math/fwd/core.hpp>
-#include <stan/math/prim/fun/square.hpp>
+#include <stan/math/fwd/fun/square.hpp>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/fwd/fun/atan2.hpp:6]: Good catch!

File: stan/math/prim/fun/asinh.hpp [hunk 2]
```diff
@@ -19,6 +19,29 @@
...
+/**
+ * Return the hyperbolic arc sine of the complex argument.
+ *
+ * @tparam V a complex type with inner arithmetic type.
+ * @param[in] x argument
+ * @return hyperbolic arc sine of the argument
+ */
+template <typename T, require_complex_bt<std::is_arithmetic, T>* = nullptr>
+inline auto asinh(const T x) {
+  return std::asinh(x);
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/prim/fun/asinh.hpp:43]: This version, with `require_complex_bt`, can only be called for `std::complex<Arithmetic>` types. So no need for autodiff here

File: stan/math/fwd/fun/hypergeometric_1F0.hpp [hunk 3]
```diff
@@ -1,9 +1,12 @@
 #ifndef STAN_MATH_FWD_FUN_HYPERGEOMETRIC_1F0_HPP
 #define STAN_MATH_FWD_FUN_HYPERGEOMETRIC_1F0_HPP
 
-#include <stan/math/prim/meta.hpp>
-#include <stan/math/prim/fun/hypergeometric_1F0.hpp>
 #include <stan/math/fwd/core.hpp>
+#include <stan/math/fwd/meta.hpp>
+#include <stan/math/rev/fun/log1m.hpp>
+#include <stan/math/rev/fun/inv.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
```
Comment (WardBrian) [inline stan/math/fwd/fun/hypergeometric_1F0.hpp:8]: Did you mean for these to be from `rev`? I assume this is a typo and should be `fwd`

File: stan/math/fwd/fun/inv_sqrt.hpp [hunk 4]
```diff
@@ -3,14 +3,14 @@
...
 #include <stan/math/fwd/core.hpp>
+#include <stan/math/prim/fun/inv_sqrt.hpp>
 #include <cmath>
 
 namespace stan {
 namespace math {
 
 template <typename T>
 inline fvar<T> inv_sqrt(const fvar<T>& x) {
-  using std::sqrt;
```
Comment (WardBrian) [inline stan/math/fwd/fun/inv_sqrt.hpp:13]: I think this file should include sqrt? or the computation should call inv_sqrt on x.val, not use sqrt?

File: stan/math/fwd/fun/log_mix.hpp [hunk 5]
```diff
@@ -28,7 +28,6 @@ template <typename T_theta, typename T_lambda1, typename T_lambda2, int N>
 inline void log_mix_partial_helper(
     const T_theta& theta, const T_lambda1& lambda1, const T_lambda2& lambda2,
     promote_args_t<T_theta, T_lambda1, T_lambda2> (&partials_array)[N]) {
-  using std::exp;
```
Comment (WardBrian) [inline stan/math/fwd/fun/log_mix.hpp:31]: This function actually calls exp, so I think we should include it from fwd?

File: stan/math/fwd/fun/tcrossprod.hpp [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
 #define STAN_MATH_FWD_FUN_TCROSSPROD_HPP
 
-#include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/fwd/meta.hpp>
+#include <stan/math/fwd/fun/multiply.hpp>
 #include <stan/math/prim/fun/to_ref.hpp>
 #include <stan/math/prim/fun/transpose.hpp>
-#include <stan/math/fwd/fun/multiply.hpp>
+#include <stan/math/prim/fun/multiply.hpp>
```
Comment (WardBrian) [inline stan/math/fwd/fun/tcrossprod.hpp:9]: Was this meant to be a deletion?

**Review by WardBrian — CHANGES_REQUESTED** [review]
A few small things in the new fwd changes, plus the above comments on doc seem to be outstanding

**Comment by WardBrian** [comment]
@SteveBronder I addressed all my own docstring comments, so it's just the ~6 things above that are unresolved

File: stan/math/fwd/fun/tcrossprod.hpp (hunk 6 above)
Comment (SteveBronder) [reply stan/math/fwd/fun/tcrossprod.hpp:9]: No sorry we are supposed to use `multiply` in this function.

multiply returns a scalar for `multiply(vector, row_vector)`. Should this function also return a scalar in the case of a vector?

File: stan/math/fwd/fun/log_mix.hpp (hunk 5 above)
Comment (SteveBronder) [reply stan/math/fwd/fun/log_mix.hpp:31]: 👍

File: stan/math/fwd/fun/inv_sqrt.hpp (hunk 4 above)
Comment (SteveBronder) [reply stan/math/fwd/fun/inv_sqrt.hpp:13]: Included `fwd/fun/sqrt.hpp`

File: stan/math/fwd/fun/tcrossprod.hpp (hunk 6 above)
Comment (SteveBronder) [reply stan/math/fwd/fun/tcrossprod.hpp]: ^yes. What about the second part of my Q. I think it should always return a matrix

File: stan/math/fwd/fun/hypergeometric_1F0.hpp (hunk 3 above)
Comment (SteveBronder) [reply stan/math/fwd/fun/hypergeometric_1F0.hpp]: 👍 typo

File: stan/math/rev/fun/hypergeometric_pFq.hpp [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 #include <stan/math/prim/meta.hpp>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/fun/hypergeometric_pFq.hpp]: need to be `/rev/meta.hpp`

File: stan/math/rev/fun/exp.hpp [hunk 8]
```diff
@@ -3,14 +3,13 @@
...
-#include <stan/math/prim/fun/exp.hpp>
+#include <stan/math/rev/fun/cos.hpp>
 #include <stan/math/rev/fun/is_inf.hpp>
-#include <stan/math/prim/fun/isinf.hpp>
-#include <stan/math/prim/fun/isfinite.hpp>
 #include <stan/math/rev/fun/is_nan.hpp>
-#include <stan/math/rev/fun/cos.hpp>
 #include <stan/math/rev/fun/sin.hpp>
+#include <stan/math/prim/fun/isinf.hpp>
+#include <stan/math/prim/fun/isfinite.hpp>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/fun/exp.hpp]: typo

File: stan/math/rev/fun/cos.hpp [hunk 9]
```diff
@@ -1,15 +1,14 @@
...
-#include <stan/math/prim/fun/cos.hpp>
-#include <stan/math/prim/fun/isinf.hpp>
-#include <stan/math/prim/fun/isfinite.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/rev/fun/abs.hpp>
 #include <stan/math/rev/fun/cosh.hpp>
 #include <stan/math/rev/fun/sinh.hpp>
+#include <stan/math/prim/fun/isinf.hpp>
+#include <stan/math/prim/fun/isfinite.hpp>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/fun/cos.hpp]: removed

File: stan/math/rev/fun/columns_dot_product.hpp [hunk 10]
```diff
@@ -2,12 +2,12 @@
...
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/err.hpp>
-#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/typedefs.hpp>
 #include <stan/math/rev/fun/dot_product.hpp>
 #include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/fun/columns_dot_product.hpp]: fixed

File: stan/math/rev/fun/asinh.hpp [hunk 11]
```diff
@@ -17,6 +12,11 @@
 #include <stan/math/rev/fun/log.hpp>
 #include <stan/math/rev/fun/polar.hpp>
 #include <stan/math/rev/fun/sqrt.hpp>
+#include <stan/math/prim/core.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/fun/asinh.hpp>
+#include <stan/math/prim/fun/isinf.hpp>
+#include <stan/math/prim/fun/is_inf.hpp>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/fun/asinh.hpp]: 👍

File: stan/math/rev/fun/atan.hpp [hunk 12]
```diff
@@ -15,6 +11,9 @@
 #include <stan/math/rev/fun/sqrt.hpp>
 #include <stan/math/rev/fun/is_inf.hpp>
 #include <stan/math/rev/fun/is_nan.hpp>
+#include <stan/math/prim/fun/isinf.hpp>
+#include <stan/math/prim/fun/isnan.hpp>
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/fun/atan.hpp]: 👍

File: stan/math/fwd/fun/log_mix.hpp (hunk 5 above)
Comment (SteveBronder) [reply stan/math/fwd/fun/log_mix.hpp:31]: 👍

File: stan/math/fwd/fun/tcrossprod.hpp (hunk 6 above)
Comment (WardBrian) [reply stan/math/fwd/fun/tcrossprod.hpp]: I think this should always return a Eigen::Matrix to match prim, rev, etc. so the latest looks correct?

**Comment by WardBrian** [comment]
@SteveBronder final check -- good to merge?

**Review by WardBrian — DISMISSED** [review]

**Review by WardBrian — APPROVED** [review]

**Comment by syclik** [comment]
@SteveBronder awesome!!!

**Pull request merged**